src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java

```diff
@@ -744,71 +756,102 @@ public void apply(final BiConsumer<String, byte[]> outputConsumer, InputTag... i
 		synchronized (this) { // guard against concurrent apply invocations
 			refresh();
 
-			if (outputBuffer == null) { // first (inputTags present) or full (no input tags) output invocation, process everything but don't output if input tags are present
-				BiConsumer<ClassInstance, byte[]> immediateOutputConsumer;
-
-				if (fixPackageAccess || hasInputTags) { // need re-processing or output buffering for repeated applies
-					outputBuffer = new ConcurrentHashMap<>();
-					immediateOutputConsumer = outputBuffer::put;
-				} else {
-					immediateOutputConsumer = (cls, data) -> outputConsumer.accept(mapClass(cls.getName()), data);
-				}
+			for (int version: mrjClasses.keySet()) {
+				mrjRefresh(version);
 
-				List<Future<?>> futures = new ArrayList<>();
+				if (outputBuffer == null) { // first (inputTags present) or full (no input tags) output invocation, process everything but don't output if input tags are present
```

Comment:
Moving outputBuffer inside the version loop doesn't make immediate sense to me. It is meant to facilitate post processing needed by package access fixing (changing the output after processing everything) or to make tagged input handling faster (input tags are for multiple inputs, e.g. jarA and jarB, that are both on the class path but need to be output separately, e.g. remapped jarA and remapped jarB, not a single jar containing the contents of A+B).

Reply:
This is intended to fix a bug when MRJ and input tag are both presents. In this case, if the output buffer is outside of the loop, and `outputBuffer` will be cleared for each mrj version, so only the last-processing mrj class file will be outputted. 

If I do not clear outputBuffer for each mrj version, then the following would happen.
> Namely, since input tag exists, outputBuffer will be used, and since outputBuffer only set to null when finish() is called, on the second time, no operation will be performed as outputBuffer == null is false. I have a test locally that can reproduce this problem. Still thinking about a solution for this.

It is possible some better solution exists for this problem. But this is the one that first came into my head.

Reply:
By the way, the `mrj4` test in the `IntegrationTest1` is specifically for this. If you try to move the `outputBuffer` inside the loop and see what happens.